ssl/ssl_lib.c

```diff
@@ -4120,13 +4209,28 @@ int ssl_validate_ct(SSL *s)
         goto end;
     }
 
+    issuer = sk_X509_value(s->verified_chain, 1);
     CT_POLICY_EVAL_CTX_set0_cert(ctx, cert);
     CT_POLICY_EVAL_CTX_set0_issuer(ctx, issuer);
     CT_POLICY_EVAL_CTX_set0_log_store(ctx, s->ctx->ctlog_store);
 
     scts = SSL_get0_peer_scts(s);
 
-    if (SCT_LIST_validate(scts, ctx) != 1) {
+    /*
+     * This function returns success (> 0) only when all the SCTs are valid, 0
+     * when some are invalid, and < 0 on various internal errors (out of
+     * memory, etc.).  Having some, or even all, invalid SCTs is not sufficient
+     * reason to abort the handshake, that decision is up to the callback.
+     * Therefore, we error out only in the unexpected case that the return
+     * value is negative.
+     *
+     * XXX: One might well argue that the return value of this function is an
+     * unforunate design choice.  Its job is only to determine the validation
+     * status of each of the provided SCTs.  So long as it correctly separates
+     * the wheat from the chaff it should return success.  Failure in this case
+     * ought to correspond to an inability to carry out its duties.
```

Comment:
The other functions have a different job and handle any callbacks internally, so that they have a final verdict by the time they return.  This is not the case here.  We're just determining the validation status of a list of SCTs, and it is in fact wrong to impute failure if just one of the elements is invalid.  I can see perhaps returning failure (0) if none are valid, but even then what to do is in the hands of the callback.

Perhaps it would be best to have the function return nothing (void), and the caller can look at the list of results and decide what to do.  If some of the SCTs didn't get a definite valid/invalid status, the caller can decide what to do.

The main thing here is that I'm fixing a bug where any single invalid SCT preempted the callback.  uite likely the callback should never be preempted.